packages/plugins/rpc-interceptors/src/lib/embark_smartContracts.ts

```diff
@@ -0,0 +1,33 @@
+import { Embark } from 'embark-core';
+import { ProxyRequestParams, ProxyResponseParams } from 'embark-proxy';
+import RpcInterceptor from "./rpcInterceptor";
+
+export default class EmbarkSmartContracts extends RpcInterceptor {
+
+  constructor(embark: Embark) {
+    super(embark);
+  }
+
+  async registerRpcInterceptors() {
+    this.embark.events.request(
+      'rpc:request:interceptor:register',
+      'embark_getSmartContracts',
+      this.interceptRequest.bind(this)
+    );
+    this.embark.events.request(
+      'rpc:response:interceptor:register',
+      'embark_getSmartContracts',
+      this.interceptResponse.bind(this)
+    );
+  }
```

Comment:
A few things we might want to consider (to improve DX):

- `registerRpcInterceptors()` will probably always involve registering a request and/or response interceptor, so there's a certain amount of boilerplate that's always the same
- Instead of requiring a `registerRpcInterceptors()` method, we could say an `RpcInterceptor` always needs an `interceptRequest(params: ProxyRequestParams<T>)` and an `interceptResponse(params: ProxyResponseParams<T, S>)` 
- Embark would then **expect** these APIs and register them as handlers for the developers
- What if an interceptor doesn't need to intercept a request or response? In such case, we would always just resolve with the same params that came in 
- What if an interceptor needs multiple interceptors for request/response? We can just have a single interceptorFn take care of multiple things

@emizzle thoughts?

Reply:
> registerRpcInterceptors() will probably always involve registering a request and/or response interceptor, so there's a certain amount of boilerplate that's always the same

Agree, except the boilerplate only exists when extending `RpcInterceptor`, and this allows the flexibility of registering the interceptions from any plugin. The `RpcInterceptor` class exists is more of a convenience class for registering Embark's core RPC interceptors, so if we wanted to remove boilerplate, we could have two methods on the `RpcInterceptor` class as you've explained.

> Instead of requiring a `registerRpcInterceptors()` method, we could say an `RpcInterceptor` always needs an `interceptRequest(params: ProxyRequestParams<T>)` and an `interceptResponse(params: ProxyResponseParams<T, S>)`

I agree with this completely. One note is that we would need to have the intercepted RPC method specified, so likely would be good to accept the method as a string parameter in the constructor, as it will be static for the lifetime of the `RpcInterceptor` class.